chronos/bipbuffer.nim

```diff
@@ -0,0 +1,167 @@
+    if spaceAfterA >= bp.a.start:
+      (spaceAfterA, bp.a.finish)
+    else:
+      (bp.a.start, Natural(0))
+
+func freeSpace*(bp: BipBuffer): Natural =
+  ## Returns amount of free space in buffer `bp`.
+  var sum: Natural
+  if len(bp.b) > 0:
+    sum += (bp.a.start - bp.b.finish)
+  else:
+    sum += bp.a.start
+  sum += Natural(len(bp.data)) - bp.a.finish
+  sum
+
+func availSpace*(bp: BipBuffer): Natural =
+  ## Returns amount of space available for reserve in buffer `bp`.
+  let (res, _) = bp.calcReserve()
+  res
+
+func usedSpace*(bp: BipBuffer): Natural =
+  ## Returns amount of used space in buffer `bp`.
+  len(bp.b) + len(bp.a)
+
+proc reserve*(bp: var BipBuffer, size: Natural): Result[Natural, cstring] =
+  ## Reserve `size` bytes in buffer. Returns number of bytes reserved.
+  doAssert(size <= len(bp.data))
+  let (availableSpace, reserveStart) = bp.calcReserve()
+  if availableSpace == 0:
+    return err("Not enough space available")
+  let reserveLength = min(availableSpace, size)
+  bp.r = BipPos.init(reserveStart, Natural(reserveStart + reserveLength))
+  ok(availableSpace)
+
+proc reserve*(bp: var BipBuffer): Natural =
+  ## Reserve all available free space in buffer. Returns number of bytes
+  ## reserved.
+  ##
+  ## Note, this procedure will raise Defect if there is no free space available.
+  let (availableSpace, reserveStart) = bp.calcReserve()
+  doAssert(availableSpace > 0, "Not enough space available")
+  bp.r = BipPos.init(reserveStart, Natural(reserveStart + availableSpace))
+  availableSpace
+
+proc reserve*(bp: var BipBuffer,
+              pt, st: typedesc): tuple[data: pt, size: st] =
+  ## Reserve all available free space in buffer. Returns current reserved range
+  ## as pointer of type `pt` and size of type `st`.
+  ##
+  ## Note, this procedure will raise Defect if there is no free space available.
+  let (availableSpace, reserveStart) = bp.calcReserve()
+  doAssert(availableSpace > 0, "Not enough space available")
+  bp.r = BipPos.init(reserveStart, Natural(reserveStart + availableSpace))
+  (cast[pt](addr bp.data[bp.r.start]), cast[st](len(bp.r)))
+
+func getReserve*(bp: var BipBuffer,
+                 pt: typedesc, st: typedesc): tuple[data: pointer, size: cint] =
+  ## Returns current reserved range as pointer + size tuple.
+  (cast[pt](addr bp.data[bp.r.start]), st(len(bp.r)))
+
+proc commit*(bp: var BipBuffer, size: Natural) =
+  ## Updates structure's pointers when new data inserted into buffer.
+  doAssert(len(bp.r) >= size,
+    "Committed size could not be larger than the previously reserved one")
+  if size == 0:
+    bp.r.zero()
+    return
+
+  let toCommit = min(size, len(bp.r))
+  if len(bp.a) == 0 and len(bp.b) == 0:
+    bp.a.start = bp.r.start
+    bp.a.finish = bp.r.start + toCommit
+  elif bp.r.start == bp.a.finish:
+    bp.a.finish += toCommit
+  else:
+    bp.b.finish += toCommit
+  bp.r.zero()
+
+proc decommit*(bp: var BipBuffer, size: Natural): Natural =
```

Comment:
boost calls these `prepare`/`commit`/`consume` - `decommit` sounds strange - also, why do you need the number of bytes consumed? 

The only way you read data from this buffer is via `bytes`/`regions` so you always know how many bytes were available, so you should in principle never `decommit` more than that.

Also, I suspect the buffer needs modified-while-iterating protection, ie an easy bug for a consumer is to call `decommit` while iterating over `regions`.

Reply:
Because there is two regions, which could cover not all the buffer length we have an iterator `bytes` and `regions`, but you still can consume/decommit less bytes than the buffer has.

Reply:
the point was more that when you're calling this function, you already know how much data you have because the code paths that lead to it require you to acquire this information and now you call consume to tell the buffer how much data you used. 

You never call this function "speculatively" to find out how much data there was in the region.

Reply:
ie the suggested action point here is to remove the return value

Reply:
(and maybe make a Defect for trying to consume more data than is in the buffer)

Reply:
I think it is already fixed, no?